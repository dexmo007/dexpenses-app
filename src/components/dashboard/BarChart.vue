<script>
import { Bar } from 'vue-chartjs';

/* eslint-disable no-underscore-dangle: "off" */
export default {
  extends: Bar,
  props: ['chartData'],
  computed: {
    options() {
      return {
        responsive: true,
        maintainAspectRatio: false,
        legend: {
          display: false,
        },
        scales: {
          xAxes: [
            {
              type: 'time',
              time: {
                unit: this.chartData.$unit,
                round: this.chartData.$unit,
                // displayFormats: {
                //   day: 'MMM D'
                // }
              },
            },
          ],
        },
        onClick: e => {
          console.log('onClick');

          // eslint-disable-next-line
          const [active] = this.$data._chart.getElementAtEvent(e);
          const target = {
            // eslint-disable-next-line
            label: active._chart.data.labels[active._index],
            // eslint-disable-next-line
            index: active._index,
          };

          this.$emit('drill', target);
        },
      };
    },
  },
  mounted() {
    this.renderChart(this.chartData, this.options);
  },
  watch: {
    chartData() {
      console.log('options changed');
      // eslint-disable-next-line
      const chart = this.$data._chart;
      chart.data = this.chartData;
      const timeOptions = chart.options.scales.xAxes[0].time;
      timeOptions.unit = this.chartData.$unit;
      timeOptions.round = this.chartData.$unit;
      // TODO: we should adapt the tooltip value for given unit
      chart.update();
    },
  },
};
</script>
