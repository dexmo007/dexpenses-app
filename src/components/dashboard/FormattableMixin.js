export default {
  props: {
    format: Object,
    locale: {
      type: String,
      default: 'de-DE'
    }
  },
  computed: {
    nf() {
      return new Intl.NumberFormat(this.locale, this.format || {})
    }
  }
}
