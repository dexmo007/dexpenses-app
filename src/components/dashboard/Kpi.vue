<template>
  <div class="kpi-container">
    <h2 v-if="title">{{title}}</h2>
    <span v-if="value || value === 0">{{value}} â‚¬</span>
    <span
      class="red--text text--lighten-1"
      v-else-if="error"
    >Error</span>
    <div
      class="loader"
      v-else
    >
      <v-progress-circular indeterminate />
    </div>
  </div>
</template>
<script>
import firebase from 'firebase/app';

export default {
  props: {
    title: String,
    func: {
      type: String,
      required: true,
    },
    data: Object,
  },
  data() {
    return {
      value: null,
      error: false,
    };
  },
  async mounted() {
    try {
      const { data } = await firebase.functions().httpsCallable(this.func)(
        this.data || {}
      );
      this.value = data.value || 0;
    } catch {
      this.error = true;
    }
  },
};
</script>
<style scoped>
.kpi-container {
  display: flex;
  flex-direction: column;
  height: 100%;
  align-items: center;
  justify-content: center;
}
.kpi-container span {
  font-weight: bold;
  font-size: 48px; /* TODO: responsive (vw?) */
}
.loader {
  height: 72px;
  display: flex;
  align-items: center;
}
</style>
