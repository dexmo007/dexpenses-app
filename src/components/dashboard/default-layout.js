export default [
  {
    x: 0,
    y: 0,
    w: 3,
    h: 2,
    i: '0',
    component: 'Kpi',
    props: { title: 'Average monthly total', func: 'aggregateAverageTotal' },
  },
  {
    x: 3,
    y: 0,
    w: 3,
    h: 2,
    i: '1',
    component: 'Kpi',
    props: {
      title: 'Current month',
      func: 'aggregateTotal',
      data: {
        start: {
          $now: {
            set: { day: 1, hour: 0, minute: 0, second: 0, millisecond: 0 },
          },
        },
      },
    },
  },
  { x: 6, y: 0, w: 3, h: 2, i: '2', component: 'TagCloud' },
  {
    x: 9,
    y: 0,
    w: 3,
    h: 2,
    i: '3',
    component: 'DoughnutAggregation',
    props: { func: 'aggregateByPaymentMethod' },
  },
  {
    x: 0,
    y: 3,
    w: 6,
    h: 3,
    i: '6',
    component: 'ExpensesOverTime',
  },
  {
    x: 6,
    y: 3,
    w: 6,
    h: 3,
    i: '7',
    component: 'Map',
    props: { center: '$home' },
  },
];
