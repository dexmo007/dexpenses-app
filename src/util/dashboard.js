import { parseProps } from './relative-dates';

// eslint-disable-next-line import/prefer-default-export
export function parseLayout(rawLayout) {
  return rawLayout.map(item => ({
    ...item,
    props: parseProps(item.props),
  }));
}
