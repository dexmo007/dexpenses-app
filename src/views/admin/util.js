import firebase from 'firebase/app';

export const categories = ['bills', 'ec', 'gas_station', 'normal', 'parking'];

export const testDataImageBucket = 'dexpenses-207219-test-images';
export const storage = firebase.app().storage(`gs://${testDataImageBucket}/`);

export function getUrl(path) {
  return `https://firebasestorage.googleapis.com/v0/b/${testDataImageBucket}/o/${encodeURIComponent(
    path
  )}?alt=media`;
}

export function buildIdentifier(info) {
  return [info.category, info.cityCode, info.name, info.classifier, info.paymentMethod]
    .filter(p => !!p)
    .join('-');
}
