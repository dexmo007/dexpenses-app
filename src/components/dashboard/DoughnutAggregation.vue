<template>
  <DoughnutChart
    v-if="chartData"
    :chart-data="chartData"
    :options="options"
  />
</template>
<script>
import firebase from 'firebase/app';
import DoughnutChart from '@/components/charts/DoughnutChart.vue';
import FormattableMixin from './FormattableMixin';

export default {
  mixins: [FormattableMixin],
  props: {
    func: String,
    data: Object,
  },
  components: { DoughnutChart },
  data() {
    const self = this;
    return {
      chartData: null,
      options: {
        responsive: true,
        maintainAspectRatio: false,
        tooltips: {
          callbacks: {
            label(tooltipItem, data) {
              const value =
                data.datasets[tooltipItem.datasetIndex].data[tooltipItem.index];
              return self.nf.format(value);
            },
          },
        },
      },
    };
  },
  computed: {
    theme() {
      return this.$vuetify.theme.isDark
        ? this.$vuetify.theme.themes.dark
        : this.$vuetify.theme.themes.light;
    },
    colors() {
      return [
        this.theme.primary,
        this.theme.error,
        this.theme.secondary,
        this.theme.success,
        this.theme.warning,
        this.theme.accent,
      ];
    },
  },
  async mounted() {
    const { data } = await firebase.functions().httpsCallable('query')({
      name: this.func,
      params: this.data || {},
    });
    this.chartData = {
      datasets: [
        {
          data: data.map(({ value }) => value),
          backgroundColor: this.colors, // TODO: maybe it is not enough
        },
      ],
      labels: data.map(({ key }) => key),
    };
  },
};
</script>
