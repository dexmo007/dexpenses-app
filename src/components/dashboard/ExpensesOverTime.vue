<template>
  <div>
    <div class="header">
      <h2>{{title}}</h2>
    </div>
    <v-breadcrumbs
      divider=">"
      :items="drillBreadcrumbItems"
    >
      <template v-slot:item="props">
        <a
          @click="jumpTo(props.item)"
          :class="[props.item.last && 'disabled']"
        >{{ props.item.text.toUpperCase() }}</a>
      </template>
    </v-breadcrumbs>
    <div
      class="chart-container"
      :class="{loading}"
    >
      <BarChart
        :style="{height: '100%', position: 'relative'}"
        :chart-data="chartData"
        v-if="chartData"
        @drill="drillDown"
        :tooltip-label="v => nf.format(v)"
      />
      <v-progress-circular
        v-if="loading"
        class="loader"
        indeterminate
        color="primary"
      ></v-progress-circular>

    </div>
  </div>
</template>
<script>
import firebase from 'firebase/app';
import { DateTime } from 'luxon';
import BarChart from '@/components/charts/BarChart.vue';
import { dateRange } from '@/util/dates';
import FormattableMixin from './FormattableMixin';

function label(period) {
  switch (period) {
    case 'year':
      return ([row]) => new Date(row.year, 0, 1);
    case 'month':
      return ([row]) => new Date(row.year, row.month - 1, 1);
    case 'day':
      return ([row]) => new Date(row.year, row.month - 1, row.day);
    case 'hour':
      return ([row]) => new Date(row.year, row.month - 1, row.day, row.hour);
    default:
      return '';
  }
}
const startEndFor = {
  month([row]) {
    return [`${row.year}-01-01`, `${row.year}-12-31`];
  },
  day([row]) {
    return [
      DateTime.fromObject({
        year: row.year,
        month: row.month,
      }).toISO(),
      DateTime.fromObject({
        year: row.year,
        month: row.month + 1,
      })
        .minus({ milliseconds: 1 })
        .toISO(),
    ];
  },
  hour([row]) {
    const day = DateTime.fromObject({
      year: row.year,
      month: row.month,
      day: row.day,
    });
    return [day.toISO(), day.plus({ days: 1, milliseconds: -1 }).toISO()];
  },
};

const breadcrumbFor = {
  month(start) {
    const [y] = start.split('-');
    return `${y}`;
  },
  day(start) {
    const [y, m] = start.split('-');
    return `${m}/${y}`;
  },
  hour(start) {
    return `${DateTime.fromISO(start).day}`;
  },
};

export default {
  mixins: [FormattableMixin],
  name: 'ExpensesOverTime',
  props: {
    title: {
      type: String,
      default: 'Expenses over time',
    },
    format: {
      type: Object,
      default() {
        return {
          style: 'currency',
          currency: 'EUR',
        };
      },
    },
  },
  components: {
    BarChart,
  },
  data() {
    return {
      loading: false,
      rows: null,
      chartData: null,
      drillLevel: 1,
      drillLevels: ['year', 'month', 'day', 'hour'],
      start: DateTime.local()
        .minus({ years: 1 })
        .toISO(),
      end: DateTime.local().toISO(),
      drillBreadcrumbs: [
        {
          text: 'Overall',
          start: '1970-01-01',
          end: DateTime.local()
            .plus({ years: 1 })
            .toISO(),
        },
        {
          text: 'Last year',
          start: DateTime.local()
            .plus({ years: -1 })
            .toISO(),
          end: DateTime.local().toISO(),
        },
      ],
    };
  },
  computed: {
    drillBreadcrumbItems() {
      return this.drillBreadcrumbs.map((bc, index, a) => ({
        ...bc,
        index,
        last: index === a.length - 1,
      }));
    },
  },
  methods: {
    async drillDown(e) {
      if (this.drillLevel === this.drillLevels.length - 1) {
        return;
      }
      this.drillLevel += 1;
      const drillSource = this.rows[e.index];
      [this.start, this.end] = startEndFor[this.drillLevels[this.drillLevel]](
        drillSource
      );
      await this.update();
      const breadcrumb = breadcrumbFor[this.drillLevels[this.drillLevel]](
        this.start,
        this.end
      );
      this.drillBreadcrumbs.push({
        text: breadcrumb,
        start: this.start,
        end: this.end,
      });
    },
    async jumpTo(level) {
      if (level.last) {
        return;
      }
      this.drillBreadcrumbs.splice(
        level.index + 1,
        this.drillBreadcrumbs.length
      );
      this.start = level.start;
      this.end = level.end;
      this.drillLevel = level.index;
      await this.update();
    },
    async update() {
      this.loading = true;
      if (!this.chartData) {
        // render dummy chart to display loading animation
        const unit = this.drillLevels[this.drillLevel];
        const from = DateTime.fromISO(this.start);
        const to = DateTime.fromISO(this.end);
        const dates = dateRange(unit, from, to);
        this.chartData = this.newChartData({
          labels: dates,
          data: Array(dates.length).fill(0),
        }); // TODO: estimate axis height?
      }
      const req = {
        period: this.drillLevels[this.drillLevel],
        start: this.start,
        end: this.end,
      };
      const result = await firebase.functions().httpsCallable('query')({
        name: 'aggregateTotalOverTimePeriod',
        params: req,
      });
      this.rows = result.data;

      this.chartData = this.newChartData({
        labels: result.data.map(label(this.drillLevels[this.drillLevel])),
        data: result.data.map(([, total]) => total),
      });
      this.loading = false;
    },
    newChartData({ labels, data }) {
      return {
        $unit: this.drillLevels[this.drillLevel],
        labels,
        datasets: [
          {
            label: 'Expenses',
            backgroundColor: (this.$vuetify.theme.isDark
              ? this.$vuetify.theme.themes.dark
              : this.$vuetify.theme.themes.light
            ).primary,
            data,
          },
        ],
      };
    },
  },
  async mounted() {
    await this.update();
  },
};
</script>
<style scoped>
.header {
  padding: 0.2em;
}
.chart-container {
  position: relative;
  height: calc(100% - 37px - 57px);
  width: 100%;
}
.chart-container.loading::before {
  content: '';
  position: absolute;
  height: 100%;
  width: 100%;
  background: white;
  opacity: 0.7;
  z-index: 50;
}
.chart-container.loading .loader {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  z-index: 51;
}
</style>
