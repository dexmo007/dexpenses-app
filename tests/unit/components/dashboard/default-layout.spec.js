import defaultLayout from '@/components/dashboard/default-layout';
import { parseLayout } from '@/util/dashboard';
import * as mockDate from 'jest-date-mock';

mockDate.advanceTo('2019-05-15T00:00:00.000Z');

jest.mock('firebase/app', () => ({
  firestore: jest.fn().mockReturnValue({
    doc: jest.fn().mockReturnValue({
      get: jest.fn().mockResolvedValue({
        exists: true,
        data() {
          return {
            homeLocation: {
              lng: 1,
              lat: 2,
            },
          };
        },
      }),
    }),
  }),
}));

jest.mock('@/store', () => ({
  state: {
    user: {
      user: { uid: 'test' },
    },
  },
}));

describe('dashboard/default-layout', () => {
  it('should be parsed correctly', async () => {
    await expect(parseLayout(defaultLayout)).resolves.toMatchSnapshot();
  });
});
