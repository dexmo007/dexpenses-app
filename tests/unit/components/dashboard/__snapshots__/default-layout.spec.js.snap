// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dashboard/default-layout should be parsed correctly 1`] = `
Array [
  Object {
    "component": "Kpi",
    "h": 2,
    "i": "0",
    "props": Object {
      "func": "aggregateAverageTotal",
      "title": "Average monthly total",
    },
    "w": 3,
    "x": 0,
    "y": 0,
  },
  Object {
    "component": "Kpi",
    "h": 2,
    "i": "1",
    "props": Object {
      "data": Object {
        "start": "2019-05-01",
      },
      "func": "aggregateTotal",
      "title": "Current month",
    },
    "w": 3,
    "x": 3,
    "y": 0,
  },
  Object {
    "component": "TagCloud",
    "h": 2,
    "i": "2",
    "props": undefined,
    "w": 3,
    "x": 6,
    "y": 0,
  },
  Object {
    "component": "DoughnutAggregation",
    "h": 2,
    "i": "3",
    "props": Object {
      "func": "aggregateByPaymentMethod",
    },
    "w": 3,
    "x": 9,
    "y": 0,
  },
  Object {
    "component": "ExpensesOverTime",
    "h": 3,
    "i": "6",
    "props": undefined,
    "w": 6,
    "x": 0,
    "y": 3,
  },
  Object {
    "component": "Map",
    "h": 3,
    "i": "7",
    "props": Object {
      "center": Object {
        "lat": 52.427547,
        "lng": 10.78042,
      },
    },
    "w": 6,
    "x": 6,
    "y": 3,
  },
]
`;
